File: rust/src/bin/step1_read_print.rs
use mal::{ast::AST, error::Error, printer, reader, readline};

fn read(string: &str) -> Result<AST, Error> {
    reader::read(string)
}

fn eval(ast: AST, _env: &str) -> Result<AST, Error> {
    Ok(ast)
}

fn print(exp: AST) -> Result<String, Error> {
    Ok(printer::print(exp))
}

fn rep(string: &str) -> Result<String, Error> {
    print(eval(read(string)?, "")?)
}

fn main() {
    let mut prompt = readline::Prompt::new();

    while let Some(line) = prompt.readline("user> ") {
        match rep(&line) {
            Ok(result) => println!("{}", result),
            Err(Error::EmptyInput) => {}
            Err(err) => eprintln!("{}", err),
        }
    }
}

File: impls/noc7c9-typescript/src/step1_read_print.ts
import * as t from './types';
import * as readline from './readline';
import * as reader from './reader';
import * as printer from './printer';

const REPL_CONTINUE = Symbol('REPL_CONTINUE');

function read(line: string): t.MalType {
    const ast = reader.readStr(line);
    if (ast == null) {
        throw REPL_CONTINUE;
    }
    return ast;
}

function eval_(ast: t.MalType): t.MalType {
    return ast;
}

function print(ast: t.MalType): string {
    return printer.printStr(ast, true);
}

(function main() {
    const prompt = readline.initialize('user> ');
    let line;
    while ((line = prompt()) != null) {
        try {
            line = read(line);
            line = eval_(line);
            line = print(line);
            console.log(line);
        } catch (err) {
            if (err === REPL_CONTINUE) continue;
            if (err instanceof Error) throw err;
            console.error('Error:', printer.printStr(err, true));
        }
    }

    process.exit(0);
})();
